### PR TITLE
adept_openstack: refactor, allow expanded operations

### PR DESCRIPTION
**Pull request by edsantiago**
Remove much duplication and dangerous hardcoded knowledge
about the names and numbers of allowed operations. End
goal is to make built-in operations more modular, allowing
for simple expanding (or removing) of operations.

Also: clarified some confusing/misleading parts.

Signed-off-by: Ed Santiago <santiago@redhat.com>

### CODE REVIEW

**Comment by cevich** [comment]
Ohh, just found this.  Silly broken github notifications :confounded:

**Comment by cevich** [comment]
Changes look fine on the surface to me, but travis is barking about some pylint stuff.  That may get fixed if you rebase and re-push.  Let me know if not, and/or you'd prefer not to to figure it out :smile: 

> Warning: completely untested

That's fine.  Adding integration tests back in is on a TODO list somewhere.  I can help test it manually (usually by running the 'basic' job).

**Comment by edsantiago** [comment]
Rebased and force-pushed. I'm baffled by pylint failing because this module never passed pylint in the first place (as in, before I touched it).

**Comment by cevich** [comment]
@edsantiago it's a "special" pylint built into the unittest, but yes, it barely passes.  I didn't investigate the issues it found here, perhaps a closer look would show that they can simply be ignored?

looking deeper...

**Comment by cevich** [comment]
@edsantiago Oh, now I see it's passing pylint, but actual unittest is failing:

```

test_operations (test_openstack.TestParsedArgs)
Check expected minimum arguments parse in predictable way ... 
======================================================================
FAIL: test_operations (test_openstack.TestParsedArgs) (expected={'flavor': 'm1.medium', 'image': 'CentOS-Cloud-7', 'private': False, 'name': 'baz'}, operation='discover', parsed_args={'debug': False, 'operation': 'discover', 'verbose': False, 'timeout': 300, 'name': 'baz'})
Check expected minimum arguments parse in predictable way
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/cevich/ADEPT/test_openstack.py", line 203, in test_operations
    self.assertDictContainsSubset(op_expected[operation], parsed_args)
AssertionError: Missing: 'flavor','image','private'
----------------------------------------------------------------------
```

Should be easy to fix.

**Comment by cevich** [comment]
BTW: U can reproduce exact same unittest run with:

``WORKSPACE=/tmp ./venv-cmd.sh unit2 -vfc``

**Comment by edsantiago** [comment]
pylint and unit tests passing now - thanks.

**Comment by cevich** [comment]
Yay!  Okay, I'll run the basic job against this...

**Comment by cevich** [comment]
@edsantiago basic job appears to be working (so ``openstack_discover_create.py`` is fine), but I found a problem with the command-line parsing for destroy and reap:

```
./openstack_destroy.py --help
ERROR:root:Unknown operation 'destro'; script must be invoked with one of the following names: "openstack_exclusive_create", "openstack_discover_create", "openstack_destroy", "openstack_reap"
usage: ./openstack_destroy.py [-h] [--verbose] [--debug] [--timeout TIMEOUT]
                              name

Low dependency script called by ADEPT playbooks to manage OpenStack VMs.
...snip...
```

That error comes from around line 1368, and relies on ``parser.exit()`` in ``parse_args()`` with the special "help" operation.  The later part appears to be happening, but clearly somewhere the ``operation`` value is getting mangled :confused:   Any ideas?

File: kommandir/bin/adept_openstack.py [hunk 1]
```diff
@@ -1357,18 +1356,17 @@ def activate_and_setup(namespace, venvdir, requirements, onlybin, nobin):
...
-        _dargs = parse_args(sys.argv, 'discover')
-    elif basename == ONLY_CREATE_NAME:
-        _dargs = parse_args(sys.argv, 'exclusive')
-    elif basename == DESTROY_NAME:
-        _dargs = parse_args(sys.argv, 'destroy')
-    elif basename == REAP_NAME:
-        _dargs = parse_args(sys.argv, 'reap')
-    else:
-        logging.error("Script was not called as %s, %s, or %s", *ALLOWED_NAMES)
+    basename = os.path.basename(sys.argv[0]).rstrip('.py')
```
Comment (cevich) [inline kommandir/bin/adept_openstack.py]: somehow(?) this rstrip seems to be eating more than just the '.py' :confused:

Comment (cevich) [reply kommandir/bin/adept_openstack.py]: my gosh that's odd...is ``rstrip()`` broken/bugged?

Comment (edsantiago) [reply kommandir/bin/adept_openstack.py]: Yeah, python strip() doesn't seem to work the way I thought it did. Looking.

Comment (cevich) [reply kommandir/bin/adept_openstack.py]: Oooohhhhh.....it's ``rstrip([CHARS])`` as in, characters to remove, not string.

Comment (cevich) [reply kommandir/bin/adept_openstack.py]: This should work:

```python
    basename = os.path.basename(sys.argv[0]).rpartition('.py')[0]
    op_requested = basename.partition('openstack_')[-1]
```

**Comment by cevich** [comment]
...basic job passed, but destruction both failed, and exited 0 :cry: That appears to be set here: https://github.com/cevich/ADEPT/pull/218/files#diff-d580606668133bebfaa51db413357274L1152

Changing that to ``parser.exit(1)`` results in non-zero exit, but the side-effect is then doing ``./openshift_whatever.py --help`` will also be non-zero exit (and maybe not a problem worth worrying about).

**Comment by cevich** [comment]
Oh, I should clarify..."but destruction failed"...using the old code (with rstrip).

File: kommandir/bin/adept_openstack.py [hunk 2]
```diff
@@ -1357,18 +1357,17 @@ def activate_and_setup(namespace, venvdir, requirements, onlybin, nobin):
...
-    elif basename == ONLY_CREATE_NAME:
-        _dargs = parse_args(sys.argv, 'exclusive')
-    elif basename == DESTROY_NAME:
-        _dargs = parse_args(sys.argv, 'destroy')
-    elif basename == REAP_NAME:
-        _dargs = parse_args(sys.argv, 'reap')
-    else:
-        logging.error("Script was not called as %s, %s, or %s", *ALLOWED_NAMES)
+    basename = re.sub(r'\.py$', '', os.path.basename(sys.argv[0]))
+    op_requested = re.sub('^openstack_', '', basename)
```
Comment (cevich) [inline kommandir/bin/adept_openstack.py]: IDK in this instance, but for future ref: ``re`` is much more "expensive" than the simpler string manip. functions.

**Comment by cevich** [comment]
...basic job running but my manual-testing all passes...

File: kommandir/bin/adept_openstack.py (hunk 1 above)
Comment (edsantiago) [reply kommandir/bin/adept_openstack.py]: New force push. Using `re`: `{,r}partition` seem close but could fail in (unlikely) corner cases. Might as well do it right. Tests pass.

File: kommandir/bin/adept_openstack.py (hunk 2 above)
Comment (edsantiago) [reply kommandir/bin/adept_openstack.py]: Python people are neurotic about re. Shrug. This is about maintainability.

Comment (cevich) [reply kommandir/bin/adept_openstack.py]: Hmmm...not sure I agree...I think it's just me that's neurotic :smile:

Comment (cevich) [reply kommandir/bin/adept_openstack.py]: gaah...github cross-comment-post.  What I mean is:  using 're' here is fine, we don't care about performance in this spot.   And you're right, it's more maintainable and less cryptic.

**Comment by cevich** [comment]
...basic job setup passed, running DaT...

File: kommandir/bin/adept_openstack.py (hunk 2 above)
Comment (edsantiago) [reply kommandir/bin/adept_openstack.py]: Ha! No, I see it all the time when looking up python answers on StackOverflow. "re is inefficient", "re is unreadable", "re is evil". It's really weird, and counterproductive. I do agree that the Python re implementation is crap, as is pretty much all Python string handling, but the extent of the loathing for re is striking.

**Comment by cevich** [comment]
...done, and all green :+1: 

Thanks @edsantiago, much cleaner code now :smile:

File: kommandir/bin/adept_openstack.py (hunk 2 above)
Comment (cevich) [reply kommandir/bin/adept_openstack.py]: > extent of the loathing for re is striking.

Interesting, I haven't encountered that.  Though I think you'll find such rantings about any language/system...people have opinions and love to spread them, especially when you disagree :smile:  But really the context definitely matters, for example right here, maintainability is _way_ more important than performance.  Best guess, it's got it's roots in 'C' (an original impetus and model for python). OTOH, try doing regex or printf stuff in the kernel, or for microcontroller and you'll be beaten to death with your own ripped-off limbs, then thrown into a volcano :laughing:
